[BUG] Downloader error getting size of GitHub zipball

**Describe the bug**

I would like to use the Downloader to get a zipball from GitHub but the execution fails. 

**To Reproduce**

Steps to reproduce the behavior:

1. Create a `power.java` file with this content:
```java
import java.nio.file.Path;

import com.onthegomap.planetiler.FeatureCollector;
import com.onthegomap.planetiler.Planetiler;
import com.onthegomap.planetiler.util.Downloader;
import com.onthegomap.planetiler.Profile;
import com.onthegomap.planetiler.config.Arguments;
import com.onthegomap.planetiler.reader.SourceFeature;

/**
 * Generates vector tiles with power lines from OpenStreetMap (see
 * https://wiki.openstreetmap.org/wiki/Tag:power%3Dline)
 * 
 * To run:
 * - java -cp planetiler.jar Power.java
 * - open power.pmtiles in https://pmtiles.io
 */
public class Power implements Profile {

  public static void main(String[] args) {
    var arguments = Arguments.fromArgs(args).withDefault("download", true);
    String area = arguments.getString("area", "geofabrik area to download", "rhode-island");
    var planetiler = Planetiler.create(arguments);

    Path dataDir = Path.of("data");
    Path sourcesDir = dataDir.resolve("sources");

    Downloader.create(planetiler.config())
      .add("pgf-encoding", "https://api.github.com/repos/wipfli/pgf-encoding/zipball/main", sourcesDir.resolve("pgf-encoding.zip")).run();


    planetiler.addOsmSource("osm", Path.of("data", area + ".osm.pbf"), "geofabrik:" + area)
        .overwriteOutput(Path.of("power.pmtiles"))
        .setProfile(new Power())
        .run();
  }

  @Override
  public void processFeature(SourceFeature sourceFeature, FeatureCollector features) {
    if (sourceFeature.canBeLine() && sourceFeature.hasTag("power", "line")) {
      features.line("power")
          .inheritAttrFromSource("voltage");
    }
  }

  @Override
  public String attribution() {
    return OSM_ATTRIBUTION;
  }

  @Override
  public boolean isOverlay() {
    return true;
  }
}
```
2. Download the latest jar `wget https://github.com/onthegomap/planetiler/releases/latest/download/planetiler.jar`
3. Run with just-in-time java 22 `java -cp planetiler.jar Power.java`
4. See error:
```
Exception in thread "main" java.lang.IllegalStateException: Error getting size of https://api.github.com/repos/wipfli/pgf-encoding/zipball/main
        at com.onthegomap.planetiler.util.Downloader.run(Downloader.java:174)
        at Power.main(Power.java:29)
Caused by: java.util.concurrent.TimeoutException
        at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
        at com.onthegomap.planetiler.util.Downloader.run(Downloader.java:168)
        at Power.main(Power.java:29)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at jdk.compiler/com.sun.tools.javac.launcher.SourceLauncher.execute(SourceLauncher.java:264)
        at jdk.compiler/com.sun.tools.javac.launcher.SourceLauncher.run(SourceLauncher.java:153)
        at jdk.compiler/com.sun.tools.javac.launcher.SourceLauncher.main(SourceLauncher.java:78)
```

**Expected behavior**
Should download the zip file.


**Environment (please complete the following information):**

- OS: Ubuntu
```
$ java --version
openjdk 22.0.1 2024-04-16
OpenJDK Runtime Environment Temurin-22.0.1+8 (build 22.0.1+8)
OpenJDK 64-Bit Server VM Temurin-22.0.1+8 (build 22.0.1+8, mixed mode, sharing)
```

--- a/planetiler-core/src/main/java/com/onthegomap/planetiler/util/Downloader.java
+++ b/planetiler-core/src/main/java/com/onthegomap/planetiler/util/Downloader.java
@@ -24,6 +24,7 @@ import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
+import java.util.OptionalLong;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
@@ -165,7 +166,7 @@ public class Downloader {
 
     for (var toDownload : toDownloadList) {
       try {
-        long size = toDownload.metadata.get(10, TimeUnit.SECONDS).size;
+        long size = toDownload.metadata.get(10, TimeUnit.SECONDS).size.orElse(0);
         loggers.addStorageRatePercentCounter(toDownload.id, size, toDownload::bytesDownloaded, true);
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
@@ -183,14 +184,19 @@ public class Downloader {
     return CompletableFuture.runAsync(RunnableThatThrows.wrap(() -> {
       LogUtil.setStage("download", resourceToDownload.id);
       long existingSize = FileUtils.size(resourceToDownload.output);
-      var metadata = httpHeadFollowRedirects(resourceToDownload.url, 0);
-      Path tmpPath = resourceToDownload.tmpPath();
-      resourceToDownload.metadata.complete(metadata);
-      if (metadata.size == existingSize) {
-        LOGGER.info("Skipping {}: {} already up-to-date", resourceToDownload.id, resourceToDownload.output);
-        return;
-      }
       try {
+        resourceToDownload.metadata.complete(httpHeadFollowRedirects(resourceToDownload.url, 0));
+      } catch (Exception e) {
+        resourceToDownload.metadata.completeExceptionally(e);
+        throw e;
+      }
+      Path tmpPath = resourceToDownload.tmpPath();
+      try {
+        var metadata = resourceToDownload.metadata.get();
+        if (metadata.size.orElse(-1) == existingSize) {
+          LOGGER.info("Skipping {}: {} already up-to-date", resourceToDownload.id, resourceToDownload.output);
+          return;
+        }
         String redirectInfo = metadata.canonicalUrl.equals(resourceToDownload.url) ? "" :
           " (redirected to " + metadata.canonicalUrl + ")";
         LOGGER.info("Downloading {}{} to {}", resourceToDownload.url, redirectInfo, resourceToDownload.output);
@@ -198,7 +204,9 @@ public class Downloader {
         FileUtils.createParentDirectories(resourceToDownload.output);
         FileUtils.delete(tmpPath);
         FileUtils.deleteOnExit(tmpPath);
-        diskSpaceCheck.addDisk(tmpPath, metadata.size, resourceToDownload.id);
+        if (metadata.size.isPresent()) {
+          diskSpaceCheck.addDisk(tmpPath, metadata.size.getAsLong(), resourceToDownload.id);
+        }
         diskSpaceCheck.checkAgainstLimits(config.force(), false);
         httpDownload(resourceToDownload, tmpPath);
         Files.move(tmpPath, resourceToDownload.output);
@@ -225,7 +233,7 @@ public class Downloader {
       responseInfo -> {
         int status = responseInfo.statusCode();
         Optional<String> location = Optional.empty();
-        long contentLength = 0;
+        OptionalLong contentLength = OptionalLong.empty();
         HttpHeaders headers = responseInfo.headers();
         if (status >= 300 && status < 400) {
           location = responseInfo.headers().firstValue(LOCATION);
@@ -235,7 +243,7 @@ public class Downloader {
         } else if (responseInfo.statusCode() != 200) {
           throw new IllegalStateException("Bad response: " + responseInfo.statusCode());
         } else {
-          contentLength = headers.firstValueAsLong(CONTENT_LENGTH).orElseThrow();
+          contentLength = headers.firstValueAsLong(CONTENT_LENGTH);
         }
         boolean supportsRangeRequest = headers.allValues(ACCEPT_RANGES).contains("bytes");
         ResourceMetadata metadata = new ResourceMetadata(location, url, contentLength, supportsRangeRequest);
@@ -250,12 +258,14 @@ public class Downloader {
     record Range(long start, long end) {}
     List<Range> chunks = new ArrayList<>();
     boolean ranges = metadata.acceptRange && config.downloadThreads() > 1;
-    long chunkSize = ranges ? chunkSizeBytes : metadata.size;
-    for (long start = 0; start < metadata.size; start += chunkSize) {
-      long end = Math.min(start + chunkSize, metadata.size);
+    boolean supportsRange = ranges && metadata.size.isPresent();
+    long fileSize = metadata.size.orElse(Long.MAX_VALUE);
+    long chunkSize = ranges ? chunkSizeBytes : fileSize;
+    for (long start = 0; start < fileSize; start += chunkSize) {
+      long end = Math.min(start + chunkSize, fileSize);
       chunks.add(new Range(start, end));
     }
-    FileUtils.setLength(tmpPath, metadata.size);
+    FileUtils.setLength(tmpPath, metadata.size.orElse(1));
     Semaphore perFileLimiter = new Semaphore(config.downloadThreads());
     Worker.joinFutures(chunks.stream().map(range -> CompletableFuture.runAsync(RunnableThatThrows.wrap(() -> {
       LogUtil.setStage("download", resource.id);
@@ -299,7 +309,7 @@ public class Downloader {
       .header(USER_AGENT, config.httpUserAgent());
   }
 
-  record ResourceMetadata(Optional<String> redirect, String canonicalUrl, long size, boolean acceptRange) {}
+  record ResourceMetadata(Optional<String> redirect, String canonicalUrl, OptionalLong size, boolean acceptRange) {}
 
   record ResourceToDownload(
     String id, String url, Path output, CompletableFuture<ResourceMetadata> metadata,

--- a/planetiler-core/src/main/java/com/onthegomap/planetiler/util/Downloader.java
+++ b/planetiler-core/src/main/java/com/onthegomap/planetiler/util/Downloader.java
@@ -258,7 +258,6 @@ public class Downloader {
     record Range(long start, long end) {}
     List<Range> chunks = new ArrayList<>();
     boolean ranges = metadata.acceptRange && config.downloadThreads() > 1;
-    boolean supportsRange = ranges && metadata.size.isPresent();
     long fileSize = metadata.size.orElse(Long.MAX_VALUE);
     long chunkSize = ranges ? chunkSizeBytes : fileSize;
     for (long start = 0; start < fileSize; start += chunkSize) {

--- a/planetiler-core/src/test/java/com/onthegomap/planetiler/util/DownloaderTest.java
+++ b/planetiler-core/src/test/java/com/onthegomap/planetiler/util/DownloaderTest.java
@@ -10,6 +10,7 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Map;
 import java.util.Optional;
+import java.util.OptionalLong;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicLong;
@@ -25,7 +26,8 @@ class DownloaderTest {
   private final PlanetilerConfig config = PlanetilerConfig.defaults();
   private AtomicLong downloads = new AtomicLong(0);
 
-  private Downloader mockDownloader(Map<String, byte[]> resources, boolean supportsRange) {
+  private Downloader mockDownloader(Map<String, byte[]> resources, boolean supportsRange,
+    boolean supportsContentLength) {
     return new Downloader(config, 2L) {
 
       @Override
@@ -55,23 +57,30 @@ class DownloaderTest {
         if (parts.length > 1) {
           int redirectNum = Integer.parseInt(parts[1]);
           String next = redirectNum <= 1 ? parts[0] : (parts[0] + "#" + (redirectNum - 1));
-          return new ResourceMetadata(Optional.of(next), url, 0, supportsRange);
+          return new ResourceMetadata(Optional.of(next), url,
+            supportsContentLength ? OptionalLong.of(0) : OptionalLong.empty(), supportsRange);
         }
         byte[] bytes = resources.get(url);
-        return new ResourceMetadata(Optional.empty(), url, bytes.length, supportsRange);
+        return new ResourceMetadata(Optional.empty(), url,
+          supportsContentLength ? OptionalLong.of(bytes.length) : OptionalLong.empty(), supportsRange);
       }
     };
   }
 
   @ParameterizedTest
   @CsvSource({
-    "false,0",
-    "true,0",
-    "false,1",
-    "false,2",
-    "true,4",
+    "false,0,true",
+    "true,0,true",
+    "false,1,true",
+    "false,2,true",
+    "true,4,true",
+
+    "false,0,false",
+    "true,0,false",
+    "false,1,false",
+    "true,1,false",
   })
-  void testDownload(boolean range, int redirects) throws Exception {
+  void testDownload(boolean range, int redirects, boolean supportsContentLength) throws Exception {
     Path dest = path.resolve("out");
     String string = "0123456789";
     String url = "http://url";
@@ -79,7 +88,7 @@ class DownloaderTest {
     Map<String, byte[]> resources = new ConcurrentHashMap<>();
 
     byte[] bytes = string.getBytes(StandardCharsets.UTF_8);
-    Downloader downloader = mockDownloader(resources, range);
+    Downloader downloader = mockDownloader(resources, range, supportsContentLength);
 
     // fails if no data
     var resource1 = new Downloader.ResourceToDownload("resource", initialUrl, dest);
@@ -96,13 +105,15 @@ class DownloaderTest {
     assertEquals(10, resource2.bytesDownloaded());
 
     // does not re-request if size is the same
-    downloads.set(0);
-    var resource3 = new Downloader.ResourceToDownload("resource", initialUrl, dest);
-    downloader.downloadIfNecessary(resource3).get();
-    assertEquals(0, downloads.get());
-    assertEquals(string, Files.readString(dest));
-    assertEquals(FileUtils.size(path), FileUtils.size(dest));
-    assertEquals(0, resource3.bytesDownloaded());
+    if (supportsContentLength) {
+      downloads.set(0);
+      var resource3 = new Downloader.ResourceToDownload("resource", initialUrl, dest);
+      downloader.downloadIfNecessary(resource3).get();
+      assertEquals(0, downloads.get());
+      assertEquals(string, Files.readString(dest));
+      assertEquals(FileUtils.size(path), FileUtils.size(dest));
+      assertEquals(0, resource3.bytesDownloaded());
+    }
 
     // does re-download if size changes
     var resource4 = new Downloader.ResourceToDownload("resource", initialUrl, dest);
@@ -131,7 +142,7 @@ class DownloaderTest {
 
       @Override
       ResourceMetadata httpHead(String url) {
-        return new ResourceMetadata(Optional.empty(), url, Long.MAX_VALUE, true);
+        return new ResourceMetadata(Optional.empty(), url, OptionalLong.of(Long.MAX_VALUE), true);
       }
     };
 

